mlir-sycl/lib/Dialect/IR/SYCLOps.cpp

```diff
@@ -98,18 +98,25 @@ mlir::LogicalResult mlir::sycl::SYCLAccessorSubscriptOp::verify() {
     return emitOpError("Dimensions cannot be zero");
 
   const auto verifyResultType = [&]() -> mlir::LogicalResult {
-    const auto resultType = getResult().getType().dyn_cast<mlir::MemRefType>();
+    const auto ResultType = getResult().getType();
+    const auto MT = ResultType.dyn_cast<mlir::MemRefType>();
+    const auto PT = ResultType.dyn_cast<LLVM::LLVMPointerType>();
 
-    if (!resultType) {
-      return emitOpError("Expecting memref return type. Got ") << resultType;
-    }
+    if (!MT && !PT)
+      return emitOpError("Expecting memref/pointer return type. Got ")
+             << ResultType;
 
-    if (resultType.getElementType() != AccessorTy.getType()) {
+    const auto ElemType = MT ? MT.getElementType() : PT.getElementType();
+    if (ElemType != AccessorTy.getType()) {
       return emitOpError(
                  "Expecting a reference to this accessor's value type (")
-             << AccessorTy.getType() << "). Got " << resultType;
+             << AccessorTy.getType() << "). Got " << ResultType;
     }
 
+    if (PT && !ElemType.isa<LLVM::LLVMStructType>())
+      return emitOpError("Expecting pointer to struct return type. Got ")
+             << ResultType;
+
```

Comment:
NIT: This LGTM, but maybe it'd be cleaner with a TypeSwitch with a single case handling both MemRef and pointer (last if would be `if constexpr (std::is_same<decltype(PT), LLVM::LLVMStructType>)` and we would not need to evaluate at runtime.

Reply:
The last if check if the element type of the llvm pointer is a struct type, we don't know the element type at compile time.

Reply:
True, we'd need to check that the input type is a pointer (at compile time) and the element type is a struct type in that case (at runtime)

Reply:
Done.